tests/test_tables/test_records.py

```diff
@@ -20,6 +23,18 @@ def test_table_records_factory(table: Table) -> None:
 
 def test_table_records_lists_factory(table: Table) -> None:
     table_records_lists = records.lists_factory(table)
+    _do_test_table_records_lists_factory(table, table_records_lists)
+
+
+def test_table_records_lists_factory_for_table_without_primary_key(metadata: MetaData) -> None:
+    table = Table('numbers', metadata, Column('number', Integer))
+    # min_size = 2 is added as a regression test for a corner-case where the
+    # absence of a primary key prevented multiple records from being generated:
+    table_records_lists = records.lists_factory(table, min_size=2)
+    _do_test_table_records_lists_factory(table, table_records_lists)
+
+
+def _do_test_table_records_lists_factory(table: Table, table_records_lists: SearchStrategy) -> None:
```

Comment:
Here, I tried to factor out the common assertion logic for re-use across both the existing randomized test and my new test. I've tried to follow the style / conventions that I've seen used in this repo, but please feel free to edit / clean up if needed.

Reply:
We can leave a single test (in the state as it was before) if we slightly improve our tests `tables` strategy (located at **tests.strategies.tables** module) to generate tables without unique columns as well like
```python
from operator import attrgetter

from hypothesis import strategies
from sqlalchemy import MetaData

from hypothesis_sqlalchemy import (columns,
                                   tables)

metadata = MetaData()
non_unique_columns = columns.non_primary_keys_factory(
        are_unique=strategies.just(False))
tables_without_unique_columns = tables.factory(
        metadatas=strategies.just(metadata),
        columns_lists=strategies.lists(non_unique_columns,
                                       unique_by=attrgetter('name')))
tables_with_unique_columns = tables.factory(
        metadatas=strategies.just(metadata))
tables = tables_without_unique_columns | tables_with_unique_columns
```

Reply:
btw this is a common scenario I use in `Hypothesis`-based tests: if there is an edge-case related to generated data that I haven't thought of, then I should rewrite my strategies to handle it (and set `repeat` option in **setup.cfg** file to some big value like `1000` and see if some error arised locally and after that rewrite tested code and be sure the problem was handled)

Reply:
This is a great suggestion.

I've been playing around with this and I've run into a surprising behavior with the proposed `tables_without_unique_columns` generator: it always seems to generate the same example when used with your `utils.example` function. For example:

```python
from tests.utils import example
print(example(tables_without_unique_columns))
print(example(tables_without_unique_columns))
```

prints

```
aaaaaaaa
aaaaaaaa
```

but

```
from tests.utils import example
print(example(tables_with_unique_columns))
print(example(tables_with_unique_columns))
```

gives me proper randomness like 

```
jfbwqyljylr
iiyebapcritaqtefhuep
```

I think this problem stems from how `example()` is defined:

```python
def example(strategy: SearchStrategy) -> Any:
    return find(specifier=strategy,
                condition=lambda x: True,
                settings=settings(phases=[Phase.generate],
                                  verbosity=Verbosity.quiet))
```

What I think happens is that `find()` tries to return a minimal example satisfying the condition, but in this new case it just so happens that every strategy's first invocation successfully generates a value (i.e. the "zero" example tried by Conjecture just-so-happened to work), whereas this wasn't the case for tables with primary keys (some constraint must be excluding the zero example).

Upstream Hypothesis's `example()` has a nice solution for this in https://github.com/HypothesisWorks/hypothesis/commit/503dbd341aecbcffd2387db0d552daa930a649e6: simply ignore the first generated example. I'm going to try changing your `example()` to simply delegate to the Hypothesis method to see if that allows my regression test to fail.

Reply:
yeah, that's true, I've ended up using this approach in newly created projects, forgot to update this one, we can use `find` function from [`topo` repo](https://github.com/lycantropos/topo/blob/master/tests/utils.py#L14)

Reply:
Update: even if I use `strategy.example()` everywhere, I'm still unable to reproduce the bug without making further test changes. I think this is because the `repeat` of `test_table_records_lists_factory` means that every invocation constructs a brand-new strategy instance (via `table_records_lists = records.lists_factory(table)`)  and then calls `.example()` on that, so we only exercise any given strategy exactly once and therefore miss out on testing more complex generated examples.

For example, the repetitions do end up ensuring that we test a variety of different table types but they never generate tables of `length > 1`.

I think that we can work around this problem by generating and testing multiple examples in each `test_table_records_lists_factory` call.

Reply:
try to rewrite `tests.utils.example` function as
```python
from typing import Any

from hypothesis import (Phase,
                        core,
                        settings)
from hypothesis.errors import (NoSuchExample,
                               Unsatisfiable)
from hypothesis.searchstrategy import SearchStrategy


def example(strategy: SearchStrategy) -> Any:
    first_object_list = []

    def condition(object_: Any) -> bool:
        if first_object_list:
            return True
        else:
            first_object_list.append(object_)
            return False

    try:
        return core.find(strategy,
                         condition,
                         settings=settings(database=None,
                                           phases=tuple(set(Phase)
                                                        - {Phase.shrink})))
    except (NoSuchExample, Unsatisfiable) as search_error:
        try:
            result, = first_object_list
        except ValueError as unpacking_error:
            raise unpacking_error from search_error
        else:
            return result
```
and let me know if it works

Reply:
Yep, that seems to work 👍 .

Out of curiosity, what's the motivation for using this instead of the built-in `example()` (whose [implementation](https://github.com/HypothesisWorks/hypothesis/blame/36f33b21b8ad2417e3c70e47c096414e17efca25/hypothesis-python/src/hypothesis/searchstrategy/strategies.py#L256) looks somewhat similar). Is there a corner-case in the built-in `example()` that your implementation avoids?

Reply:
Sorry for ignoring last question, answer is simple: mentioned [change](https://github.com/HypothesisWorks/hypothesis/commit/503dbd341aecbcffd2387db0d552daa930a649e6) wasn't there at earlier versions and I've wasted large amount of time figuring it out, so after that I've decided to write my own version and support it independently from `example` method